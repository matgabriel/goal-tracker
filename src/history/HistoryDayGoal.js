// Journée historique d’un objectif
// ================================

// Sous-section de l'écran d’historique, dédiée à un objectif sur un jour
// précis.  Rendu en fait à l'intérieur d'un autre composant bête: `HistoryDay`.

import React from 'react'

import ListItem from '@material-ui/core/ListItem'
import ListItemText from '@material-ui/core/ListItemText'

import Gauge from '../shared/Gauge'
import {
  GoalPropType,
  HistoryDayProgressesPropType,
} from '../shared/prop-types'

// La déstructuration et les valeurs par défaut en force !
const HistoryDayGoal = ({
  goal: { name, units },
  stats: [progress = 0, target] = [],
}) => {
  const details = (
    <div>
      <Gauge value={progress} max={target} />
      {progress} {units} sur {target}
    </div>
  )

  return (
    <ListItem>
      <ListItemText
        component='div'
        primary={name}
        secondary={details}
        secondaryTypographyProps={{ component: 'div' }}
      />
    </ListItem>
  )
}

// Comme toujours, on définit les propriétés attendues/autorisées pour
// validation.
HistoryDayGoal.propTypes = {
  goal: GoalPropType.isRequired,
  stats: HistoryDayProgressesPropType.isRequired,
}

export default HistoryDayGoal
