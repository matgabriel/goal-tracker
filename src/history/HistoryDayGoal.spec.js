import React from 'react'
import { render } from '@testing-library/react'

import HistoryDayGoal from './HistoryDayGoal'

describe('<HistoryDayGoal />', () => {
  const goal = { id: 0, name: 'Demo goal', target: 5, units: 'wombats' }

  it('should use proper stats propTypes', () => {
    checkFailPropType('stats in HistoryDayGoal must be an array.', () => (
      <HistoryDayGoal goal={goal} />
    ))
    checkFailPropType(
      'stats in HistoryDayGoal must be a pair of integers.',
      () => <HistoryDayGoal goal={goal} stats={[42]} />
    )
    checkFailPropType(
      'stats in HistoryDayGoal must start with a non-negative progress value.',
      () => <HistoryDayGoal goal={goal} stats={[-1, 10]} />
    )
    checkFailPropType(
      'stats in HistoryDayGoal must end with a positive target value.',
      () => <HistoryDayGoal goal={goal} stats={[0, 0]} />
    )
  })

  it('should otherwise match the expected snapshot', () => {
    const { container } = render(<HistoryDayGoal goal={goal} stats={[2, 5]} />)

    expect(container).toMatchSnapshot()
  })
})

function checkFailPropType(message, block) {
  const cErr = jest.spyOn(global.console, 'error').mockImplementation(() => {})
  try {
    block()
    expect(cErr).toHaveBeenCalledWith(
      expect.stringMatching(
        new RegExp(
          `Failed prop type: ${message.replace(/[\].*+?(){}[]/g, '\\$&')}`
        )
      )
    )
  } finally {
    cErr.mockRestore()
  }
}
