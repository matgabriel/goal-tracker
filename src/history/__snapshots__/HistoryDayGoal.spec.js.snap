// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistoryDayGoal /> should otherwise match the expected snapshot 1`] = `
<div>
  <li
    class="MuiListItem-root MuiListItem-gutters"
  >
    <div
      class="MuiListItemText-root MuiListItemText-multiline"
      component="div"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1"
      >
        Demo goal
      </span>
      <div
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary"
      >
        <div>
          <div
            aria-valuenow="40"
            class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate"
            role="progressbar"
            style="height: 8px;"
          >
            <div
              class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate"
              style="transform: translateX(-60%);"
            />
          </div>
          2
           
          wombats
           sur 
          5
        </div>
      </div>
    </div>
  </li>
</div>
`;
